src/awkward/_namedaxis.py

```diff
@@ -0,0 +1,795 @@
+    If the axis name is a valid Python identifier, it is represented as is.
+    Otherwise, it is represented as a JSON string.
+
+    Args:
+        named_axis (AxisMapping): The named axis mapping to prettify.
+
+    Returns:
+        str: The prettified string representation of the named axis mapping.
+
+    Examples:
+        >>> _prettify_named_axes({"x": 0, "y": 1, "z": 2})
+        'x:0, y:1, z:2'
+        >>> _prettify_named_axes({"x": 0, "y": 1, "$": 2})
+        'x:0, y:1, "$":2'
+    """
+
+    def _prettify(ax: AxisName) -> str:
+        repr_ax = str(ax)
+        if re.match("[A-Za-z_][A-Za-z_0-9]*", repr_ax):
+            return repr_ax
+        return json.dumps(repr_ax)
+
+    sorted_named_axis = sorted(named_axis.items(), key=lambda x: x[1])
+    return ", ".join(
+        [f"{_prettify(named_ax)}:{pos_ax}" for named_ax, pos_ax in sorted_named_axis]
+    )
+
+
+def _neg2pos_axis(
+    axis: int,
+    total: int,
+) -> int:
+    """
+    Converts a negative axis index to a positive one.
+
+    This function takes a negative axis index and the total number of axes and returns the corresponding positive axis index.
+    If the input axis index is already positive, it is returned as is.
+
+    Args:
+        axis (int): The axis index to convert. Can be negative.
+        total (int): The total number of axes.
+
+    Returns:
+        int: The positive axis index corresponding to the input axis index.
+
+    Examples:
+        >>> _neg2pos_axis(-1, 3)
+        2
+        >>> _neg2pos_axis(1, 3)
+        1
+    """
+    if axis < 0:
+        return total + axis
+    return axis
+
+
+def _get_named_axis(
+    ctx: MaybeSupportsNamedAxis | AttrsNamedAxisMapping | tp.Mapping | tp.Any,
+) -> AxisMapping:
+    """
+    Retrieves the named axis from the provided context. The context can either be an object that supports named axis
+    or a dictionary that includes a named axis mapping.
+
+    Args:
+        ctx (MaybeSupportsNamedAxis | AttrsNamedAxisMapping | Mapping | Any): The context from which the named axis is to be retrieved.
+
+    Returns:
+        AxisMapping: The named axis retrieved from the context. If the context does not include a named axis,
+            an empty dictionary is returned.
+
+    Examples:
+        >>> class Test(MaybeSupportsNamedAxis):
+        ...     @property
+        ...     def attrs(self):
+        ...         return {NAMED_AXIS_KEY: {"x": 0, "y": 1, "z": 2}}
+        ...
+        >>> _get_named_axis(Test())
+        {"x": 0, "y": 1, "z": 2}
+        >>> _get_named_axis({NAMED_AXIS_KEY: {"x": 0, "y": 1, "z": 2}})
+        {"x": 0, "y": 1, "z": 2}
+        >>> _get_named_axis({"other_key": "other_value"})
+        {}
+    """
+    if isinstance(ctx, MaybeSupportsNamedAxis):
+        return _get_named_axis(ctx.attrs)
+    elif isinstance(ctx, tp.Mapping) and NAMED_AXIS_KEY in ctx:
+        return dict(ctx[NAMED_AXIS_KEY])
+    else:
+        return {}
```

Comment:
I'm tempted to say that this function is doing too much; it can be simplified to 
```python
_get_named_axis(array.attrs)
```

It's not a big deal either way, by my instinct is that the high-level callers know whether they're passing in `ctx` or `ctx.attrs`, so they can avoid hitting this if branch altogether.

Reply:
Yes, but I wanted it to also work with objects (e.g. `HighLevelContexts` or `np.arrays`) that don't have a `.attrs` attribute without me worrying about it. In principle this is similar to `attrs_of_obj`, but this one does not work with `HighLevelContext` (and I couldn't make it work properly with it).